### ISSUE TITLE
Create new Param annotation to work with ParamFetcher

### ISSUE THREAD

**Issue by ncrypthic** [FriendsOfSymfony/FOSRestBundle#1075]
Hi,

I want to create a new annotation to pass request header information to ParamFetcher. While digging into this I found out the [```ParamFetcher```](https://github.com/FriendsOfSymfony/FOSRestBundle/blob/master/Request/ParamFetcher.php#L126-L132) was hard coded to operate only when the param is a ```QueryParam``` or ```RequestParam``` instance.

Can we improve this to utilize some interface (e.g ParamInterface) to get the value from callback with Request object passed as it parameter?

```php
<?php
namespace FOS\RestBundle\Util;

use Symfony\Component\HttpFoundation\Request;

interface ParamInteface {
    public function getValue(Request $request);
}
```

```php
<?php
namespace FOS\RestBundle\Controller\Annotations;

abstract class Param implements ParamInterface
{
}
```

```php
<?php
namespace FOS\RestBundle\Request;

use FOS\RestBundle\Util\ParamInterface;
// ...
class ParamFetcher implements ParamFetcherInterface {
    // ...
    /**
    * {@inheritDoc}
    */
    public function get($name, $strict = null)
    {
        // ...
        if ($config instanceof ParamInterface) {
            $param = $config->getValue($this->request);
        } else {
            $param = null;
        }
        // ...
    }
}
```

If this is acceptable, I can make a PR of this.

Cheers,
Lim

**Comment by lsmith77** [comment]
I guess moving the logic into the Param classes makes sense, but I would want to do this in a separate PR that we can put into 2.0 ..

Support for headers is something we could consider for 1.8 .. though maybe in the end the next step after 1.7 will be 2.0 anyway.

**Comment by ncrypthic** [comment]
This is how I do this on current version:

- I create a new kernel.controller event listener
- On the event method I get ParamFetcher service instance and set the parameter manually

Thats why I propose above implementation. If you decide to implement this, just tell me if there's anything I can help with.

**Comment by Ener-Getick** [comment]
+1

I am wondering if we should not create a ``getConstraints`` method in this interface to simplify the ``ParamFetcher``.

**Comment by lsmith77** [comment]
sounds good to me

**Comment by lsmith77** [comment]
#1097 was merged

**Comment by Ener-Getick** [comment]
This feature will be available in 2.0.
